### PR TITLE
feat(node): v13.11

### PR DESCRIPTION
**Pull request by SimonSchick**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/nodejs/node/blob/master/doc/changelogs/CHANGELOG_V13.md
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`. If for reason the any rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]` and not for whole package so that the need for disabling can be reviewed.

I specifically did NOT implement the pipeline changes as varargs in the middle of a function are not possible to be mapped in JS/TS.

@Flarna for early 👀

### CODE REVIEW

File: types/node/fs.d.ts [hunk 1]
```diff
@@ -802,13 +803,13 @@ declare module "fs" {
      * @param options Either the file mode, or an object optionally specifying the file mode and whether parent folders
      * should be created. If a string is passed, it is parsed as an octal integer. If not specified, defaults to `0o777`.
      */
-    function mkdir(path: PathLike, options: number | string | MakeDirectoryOptions | undefined | null, callback: NoParamCallback): void;
+    function mkdir(path: PathLike, options: number | string | MakeDirectoryOptions | undefined | null, callback: NoParamCallback): void | string;
```
Comment (ExE-Boss) [inline types/node/fs.d.ts]:
```suggestion
    function mkdir(path: PathLike, options: number | string | MakeDirectoryOptions | undefined | null, callback: (err: NodeJS.ErrnoException | null, path: string | undefined) => void): void;
```

File: types/node/fs.d.ts [hunk 2]
```diff
@@ -802,13 +803,13 @@ declare module "fs" {
...
 
     /**
      * Asynchronous mkdir(2) - create a directory with a mode of `0o777`.
      * @param path A path to a file. If a URL is provided, it must use the `file:` protocol.
      */
-    function mkdir(path: PathLike, callback: NoParamCallback): void;
+    function mkdir(path: PathLike, callback: NoParamCallback): void | string;
```
Comment (ExE-Boss) [inline types/node/fs.d.ts]:
```suggestion
    function mkdir(path: PathLike, callback: NoParamCallback): void;
```

File: types/node/fs.d.ts [hunk 3]
```diff
@@ -827,7 +828,7 @@ declare module "fs" {
      * @param options Either the file mode, or an object optionally specifying the file mode and whether parent folders
      * should be created. If a string is passed, it is parsed as an octal integer. If not specified, defaults to `0o777`.
      */
-    function mkdirSync(path: PathLike, options?: number | string | MakeDirectoryOptions | null): void;
+    function mkdirSync(path: PathLike, options?: number | string | MakeDirectoryOptions | null): void | string;
```
Comment (ExE-Boss) [inline types/node/fs.d.ts]:
```suggestion
    function mkdirSync(path: PathLike, options?: number | string | MakeDirectoryOptions | null): string | undefined;
```

**Review by ExE-Boss — CHANGES_REQUESTED** [review]
There should probably also be overloads that take `options: MakeDirectoryOptions & { recursive: true }`.

---

Also, generic rest parameters require **TypeScript 3.0**, which is why the build is failing.

File: types/node/test/vm.ts [hunk 4]
```diff
@@ -60,3 +60,12 @@ import { inspect } from 'util';
         cachedData: Buffer.from('nope'),
     });
 }
+
+{
+    const usage = measureMemory({
+        mode: 'detailed',
+        context: createContext(),
+    }).then((data: MemoryMeasurement) => {
+
```
Comment (ExE-Boss) [inline types/node/test/vm.ts]: **TSLint** complains that there’s an empty line here:

```suggestion
        data; // $ExpectType MemoryMeasurement
```

**Comment by SimonSchick** [comment]
I know I should've run the tests locally, was in a hurry 😛

**Comment by SimonSchick** [comment]
@ExE-Boss suggestions for avoid the generic?
Just making it `any[]` for now?
I'm not exactly sure why these functions allow passing values as a closure can just capture local variables, I doubt many people will use the parameter list.

File: types/node/fs.d.ts [hunk 5]
```diff
@@ -784,31 +784,31 @@ declare module "fs" {
...
     }
 
+    interface MakeDirectoryOptionsRecursive extends MakeDirectoryOptions {
+        /**
+         * Indicates whether parent folders should be created.
+         * If a folder was created, the path to the first created folder will be returned.
+         * @default false
+         */
+        recursive?: boolean;
+    }
```
Comment (ExE-Boss) [inline types/node/fs.d.ts]:
```suggestion
        /**
         * Indicates whether parent folders should be created.
         * If a folder was created, the path to the first created folder will be returned.
         * @default false
         */
        recursive?: boolean;

        /**
         * A file mode. If a string is passed, it is parsed as an octal integer. If not specified
         * @default 0o777.
         */
        mode?: number;
    }
```

File: types/node/fs.d.ts [hunk 6]
```diff
@@ -784,31 +784,31 @@ declare module "fs" {
...
+
     /**
      * Asynchronous mkdir(2) - create a directory.
      * @param path A path to a file. If a URL is provided, it must use the `file:` protocol.
      * @param options Either the file mode, or an object optionally specifying the file mode and whether parent folders
      * should be created. If a string is passed, it is parsed as an octal integer. If not specified, defaults to `0o777`.
      */
+    function mkdir(path: PathLike, callback: (err: NodeJS.ErrnoException, path: string | undefined) => void): void;
     function mkdir(path: PathLike, options: number | string | MakeDirectoryOptions | undefined | null, callback: NoParamCallback): void;
```
Comment (ExE-Boss) [inline types/node/fs.d.ts]:
```suggestion
    function mkdir(path: PathLike, options: MakeDirectoryOptions & { recursive: true }, callback: (err: NodeJS.ErrnoException | null, path: string) => void): void;

    /**
     * Asynchronous mkdir(2) - create a directory.
     * @param path A path to a file. If a URL is provided, it must use the `file:` protocol.
     * @param options Either the file mode, or an object optionally specifying the file mode and whether parent folders
     * should be created. If a string is passed, it is parsed as an octal integer. If not specified, defaults to `0o777`.
     */
    function mkdir(path: PathLike, options: number | string | (MakeDirectoryOptions & { recursive?: false; }) | null | undefined, callback: NoParamCallback): void;

    /**
     * Asynchronous mkdir(2) - create a directory.
     * @param path A path to a file. If a URL is provided, it must use the `file:` protocol.
     * @param options Either the file mode, or an object optionally specifying the file mode and whether parent folders
     * should be created. If a string is passed, it is parsed as an octal integer. If not specified, defaults to `0o777`.
     */
    function mkdir(path: PathLike, options: number | string | MakeDirectoryOptions | null | undefined, callback: (err: NodeJS.ErrnoException | null, path: string | undefined) => void): void;
```

File: types/node/fs.d.ts [hunk 7]
```diff
@@ -784,31 +784,31 @@ declare module "fs" {
...
-
-    /**
-     * Asynchronous mkdir(2) - create a directory with a mode of `0o777`.
-     * @param path A path to a file. If a URL is provided, it must use the `file:` protocol.
-     */
-    function mkdir(path: PathLike, callback: NoParamCallback): void;
+    function mkdir(path: PathLike, options: MakeDirectoryOptionsRecursive, callback: (err: NodeJS.ErrnoException, path: string | undefined) => void): void;
```
Comment (ExE-Boss) [inline types/node/fs.d.ts]:
```suggestion

    /**
     * Asynchronous mkdir(2) - create a directory with a mode of `0o777`.
     * @param path A path to a file. If a URL is provided, it must use the `file:` protocol.
     */
    function mkdir(path: PathLike, callback: NoParamCallback): void;
```

File: types/node/fs.d.ts [hunk 8]
```diff
@@ -819,6 +819,7 @@ declare module "fs" {
          * should be created. If a string is passed, it is parsed as an octal integer. If not specified, defaults to `0o777`.
          */
         function __promisify__(path: PathLike, options?: number | string | MakeDirectoryOptions | null): Promise<void>;
+        function __promisify__(path: PathLike, options: MakeDirectoryOptionsRecursive): Promise<undefined | string>;
```
Comment (ExE-Boss) [inline types/node/fs.d.ts]:
```suggestion
        function __promisify__(path: PathLike, options: MakeDirectoryOptions & { recursive: true; }): Promise<string>;

        /**
         * Asynchronous mkdir(2) - create a directory.
         * @param path A path to a file. If a URL is provided, it must use the `file:` protocol.
         * @param options Either the file mode, or an object optionally specifying the file mode and whether parent folders
         * should be created. If a string is passed, it is parsed as an octal integer. If not specified, defaults to `0o777`.
         */
        function __promisify__(path: PathLike, options?: number | string | (MakeDirectoryOptions & { recursive?: false; }) | null): Promise<void>;

        /**
         * Asynchronous mkdir(2) - create a directory.
         * @param path A path to a file. If a URL is provided, it must use the `file:` protocol.
         * @param options Either the file mode, or an object optionally specifying the file mode and whether parent folders
         * should be created. If a string is passed, it is parsed as an octal integer. If not specified, defaults to `0o777`.
         */
        function __promisify__(path: PathLike, options?: number | string | MakeDirectoryOptions | null): Promise<string | undefined>;
```

File: types/node/fs.d.ts [hunk 9]
```diff
@@ -828,6 +829,7 @@ declare module "fs" {
      * should be created. If a string is passed, it is parsed as an octal integer. If not specified, defaults to `0o777`.
      */
     function mkdirSync(path: PathLike, options?: number | string | MakeDirectoryOptions | null): void;
+    function mkdirSync(path: PathLike, options: MakeDirectoryOptionsRecursive): undefined | string;
```
Comment (ExE-Boss) [inline types/node/fs.d.ts]:
```suggestion
    function mkdirSync(path: PathLike, options: MakeDirectoryOptions & { recursive: true; }): string;

    /**
     * Synchronous mkdir(2) - create a directory.
     * @param path A path to a file. If a URL is provided, it must use the `file:` protocol.
     * @param options Either the file mode, or an object optionally specifying the file mode and whether parent folders
     * should be created. If a string is passed, it is parsed as an octal integer. If not specified, defaults to `0o777`.
     */
    function mkdirSync(path: PathLike, options?: number | string | (MakeDirectoryOptions & { recursive?: false; }) | null): void;

    /**
     * Synchronous mkdir(2) - create a directory.
     * @param path A path to a file. If a URL is provided, it must use the `file:` protocol.
     * @param options Either the file mode, or an object optionally specifying the file mode and whether parent folders
     * should be created. If a string is passed, it is parsed as an octal integer. If not specified, defaults to `0o777`.
     */
    function mkdirSync(path: PathLike, options?: number | string | MakeDirectoryOptions | null): string | undefined;
```

**Review by ExE-Boss — CHANGES_REQUESTED** [review]
The generic `...args` overloads of `AsyncLocalStorage` methods should be added to `types/node/ts3.1/asnyc_hooks.d.ts`.

Also, the `fs.mkdir(…)` recursive method definitions are currently incorrect.

**Pull request closed without merging**